### ISSUE TITLE
Parser recursive calls and my new match fields in Disjunct_struct

### ISSUE THREAD

**Issue by ampli** [opencog/link-grammar#244]
Hi Linas,
I implemented the idea to move connector matching to form_match_list() and use 2 new fields in the Disjunct_struct to indicate left/righ connector matching, as I mentioned in #219.
```
struct Disjunct_struct
{
    Disjunct *next;
    const char * string;       /* subscripted dictionary word */
    Connector *left, *right;
    double cost;
    bool marked;               /* unmarked disjuncts get deleted */
    bool match_left, match_right; /* used only during parsing */       <--- Added
    const Gword **word;        /* NULL terminated list of originating words */
};
```
However, I have a question regarding the safety of this idea.

My said implementation allows several shortcuts in connector matching, and speeds up en/corpus-fix-long.batch by about 40%.  To be sure it doesn't introduce any problem, I checked the detailed run of en/corpus-basic.batch, en/corpus-fixes.batch, ru/corpus-basic.batch and also he/corpus-basic.batch with limit=10000, and the results were totally identical to those of the current master branch.

However, due to the recursive working of the parser, new form_match_lists() calls are issued while disjunct list from previous ones have not been completely handled. So if the same disjuncts are **ever** happened to be touched, havoc may happen, as the said match fields would get overwrite, possibly with different results.

So I used a debug check to validate that the match list disjunct fields are not modified in the recursive calls. And indeed they are never modified by these calls. I don't yet understand the parsing algorithm, but from inspection of the code it seems this is a property of the algorithm that the same `w` is not used again in the recursive calls.

**Question**:  Can we depend on this to always be true?

**Comment by linas** [comment]
Heh.  There are three ways to answer this question:
1)  for you to fully understand the parse algo.
2) to perform a statistically test/analysis
3) for me to fully understand the algo, and for you to ask me.

Seems like you opted for 2 & 3. :-)  For me to answer 3), I would have to look what you did, study and ponder for at least a few hours or a day.  Which is time I don't have right now.  My intuition is that your study (2) is sufficient, assuming you did not make any mistakes.

Perhaps it is time to take a shot at (1)  Its actually not that hard, *if* you read the paper(s) first.  They present pseudocode which is relatively easy to understand (easier than the code; they actually explain it), and you can then compare with the actual C code, and it matches up well.  

One of the early papers gives a short pseudocode, and another gives a longer more detailed one.  The longer one is this:  Dennis Grinberg, John Lafferty, Daniel Sleator, "A Robust Parsing Algorithm for Link Grammar", Carnegie Mellon University Computer Science technical report CMU-CS-95-125, and Proceedings of the Fourth International Workshop on Parsing Technologies, Prague (1995) http://www.cs.cmu.edu/afs/cs.cmu.edu/project/link/pub/www/papers/ps/tr95-125.pdf and I think there is a shorter one here, maybe: http://www.cs.cmu.edu/afs/cs.cmu.edu/project/link/pub/www/papers/ps/tr91-196.pdf  yeah, page 29

**Comment by ampli** [comment]
There is a big question, supposing I will understand the algorithm, whether I can prove that this apparent behavior, of never touching the same disjuncts in the recursive calls, is a feature of the algorithm, as I am generally weak in proving such things...
So I will try to read the papers you mentioned, but there is no much chance (to say the least) that I will be able to prove that.

What I can do is to leave the said test under #ifdef DEBUG, as it involves only setting a field in the match-lists disjuncts (in form_match_list()) to a number which is increased in each call, and checking in do_count() that it is unchanged while it traverses the match-list.

I also cannot prove that the wordgraph flattening algorithm is correct. Hence I have put  several assert() calls in critical places to detect if anything goes wrong. And just today I found a bug in it (that triggered the "Extra word" assert()) that I am now trying to fix...

So how to proceed in this form_match_list() improvement, taking into account the extremely  reasonable assumption that I will not be able to prove such a feature of the algorithm? This function is a hotspot section in the parser, and this speed up together with the previous one already sped it up by 60% on the fix-long batch... and there is very much room there for further speed improvements!

**Comment by linas** [comment]
well, provability has always been difficult in any kind of code.  This is why assert() is used. It is also why code reviews take place. .. and why testing is done.  No one technique is ever quite enough.

If you can design and code this in some sort of way that makes it easy to review, then I can look at it .. maybe not for a week or two. or three or maybe a lot sooner, it depends.   Make the changes minimal, so that its clearly apparent how things changed.

FWIW, in the past, I've found a few bugs in LG, where it would parse wikipedia just fine, for weeks in a row on multiple cores, and then crash with a bug.  So .. sometimes bugs are rare. 

In your case, perhaps you could add one more boolean flag, a "data is currently valid" flag, that is set when you first set the data into the disjunct, and reset after the data has been "picked up".  That way, the setter can check if its OK to set the data, and the receiver can check if its receiving stale data a second time (or if it hasn't been set yet)   Another possibility is to do this, but use some sort of counter, incrementing by 1 every time you set the data (and maybe incrementing again when you receive it, so you get a even/odd sequence.)

**Comment by ampli** [comment]
>If you can design and code this in some sort of way that makes it easy to review, then I can look at it ..

Fortunately I already took this into account and wrote it this way. The code uses, as processing shortcuts, several assumptions on the features of connector strings and disjuncts, so in order to ease the review I wrote it as a series of small and hopefully clear "evolutionary" changes, each with its own detailed commit message. The commit messages include statistics and benchmarks.
